packages/common/src/ContractInteractor.ts

```diff
@@ -586,9 +633,112 @@ export class ContractInteractor {
+    _: {
+      msgData: PrefixedHexString
+      gasAndDataLimits: PaymasterGasAndDataLimits
+      relayCallGasLimit: string
+    }): number {
+    const msgDataLength = toBuffer(_.msgData).length
+    const msgDataGasCostInsideTransaction =
+      new BN(this.environment.dataOnChainHandlingGasCostPerByte)
+        .muln(msgDataLength)
+        .toNumber()
+    const calldataCost = this.calculateCalldataCost(_.msgData)
+    const result = parseInt(this.relayHubConfiguration.gasOverhead.toString()) +
+      msgDataGasCostInsideTransaction +
+      calldataCost +
+      parseInt(_.relayCallGasLimit) +
+      parseInt(_.gasAndDataLimits.preRelayedCallGasLimit.toString()) +
+      parseInt(_.gasAndDataLimits.postRelayedCallGasLimit.toString())
+    this.logger.debug(`
+input:\n${JSON.stringify(_)}
+msgDataLength: ${msgDataLength}
+calldataCost: ${calldataCost}
+msgDataGasCostInsideTransaction: ${msgDataGasCostInsideTransaction}
+environment: ${JSON.stringify(this.environment)}
+relayHubConfiguration: ${JSON.stringify(this.relayHubConfiguration)}
+calculateTransactionMaxPossibleGas: result: ${result}
+`)
+    return result
+  }
+
+  /**
+   * @param relayCallABI actual inputs of the 'relayCall' method
+   * @return {PrefixedHexString} exact calculation of how much gas sending this data will consume
+   */
+  calculateCalldataCostAbi (relayCallABI: RelayCallABI): PrefixedHexString {
+    const encodedData = this.encodeABI({ ...relayCallABI })
+    return `0x${this.calculateCalldataCost(encodedData).toString(16)}`
+  }
+
+  /**
+   * @param relayRequestOriginal request input of the 'relayCall' method with some fields not yet initialized
+   * @param variableFieldSizes configurable sizes of 'relayCall' parameters with variable size types
+   * @return {PrefixedHexString} top boundary estimation of how much gas sending this data will consume
+   */
+  estimateCalldataCostForRequest (
+    relayRequestOriginal: RelayRequest,
+    variableFieldSizes: { maxApprovalDataLength: number, maxPaymasterDataLength: number }
+  ): PrefixedHexString {
+    // protecting the original object from temporary modifications done here
+    const relayRequest: RelayRequest =
+      Object.assign(
+        {}, relayRequestOriginal,
+        {
+          relayData: Object.assign({}, relayRequestOriginal.relayData)
+        })
+    relayRequest.relayData.transactionCalldataGasUsed = '0xffffffffff'
+    relayRequest.relayData.paymasterData = '0x' + 'ff'.repeat(variableFieldSizes.maxPaymasterDataLength)
+    const maxAcceptanceBudget = constants.MAX_UINT256.toString()
+    const signature = '0x' + 'ff'.repeat(65)
+    const approvalData = '0x' + 'ff'.repeat(variableFieldSizes.maxApprovalDataLength)
+    const encodedData = this.encodeABI({
+      relayRequest,
+      signature,
+      approvalData,
+      maxAcceptanceBudget
+    })
+    return `0x${this.calculateCalldataCost(encodedData).toString(16)}`
+  }
+
+  calculateCalldataCost (msgData: PrefixedHexString): number {
```

Comment:
This method is moved in from there. `Utils` is not a class so doesn't have an `environment` field so it would have to be passed in as a parameter, so there is little point of having it in `Utils`.
In tests it was only needed in a very few places that actually care about making a correct charge.